File: 0670-maximum-swap/0670-maximum-swap.java
class Solution {
    public int maximumSwap(int num) {
        var numStr = String.valueOf(num).toCharArray();
        var maxFromRight = new int[numStr.length];
        maxFromRight[maxFromRight.length - 1] = numStr.length - 1;
        for(var right = maxFromRight.length - 2; right >= 0; right--) {
            if(numStr[right] - 'a' <= numStr[maxFromRight[right + 1]] - 'a') {
                maxFromRight[right] = maxFromRight[right + 1];
            } else {
                maxFromRight[right] = right;
            }
        }

        for(var i = 0; i < numStr.length; i++) {
            if(numStr[i] < numStr[maxFromRight[i]]) {
                var tmp = numStr[i];
                numStr[i] = numStr[maxFromRight[i]];
                numStr[maxFromRight[i]] = tmp;
                break;
            }
        }

        return Integer.parseInt(new String(numStr));
    }
}

File: 0670-maximum-swap/0670-maximum-swap.cpp
class Solution {
public:
    int maximumSwap(int num) {
        string numStr = to_string(num);
        int maxValueIndex = numStr.size() - 1;
        int swap1 = -1, swap2 = maxValueIndex;

        for(int i = numStr.size() - 2; i >= 0; i--) {
          if(numStr[i] > numStr[maxValueIndex]) {
            maxValueIndex = i;
          } else if(numStr[i] < numStr[maxValueIndex]) {
            swap1 = i;
            swap2 = maxValueIndex;
          }
        }

        if(swap1 != -1) {
            char tmp = numStr[swap1];
            numStr[swap1] = numStr[swap2];
            numStr[swap2] = tmp;
        }

        return stoi(numStr);
    }
};